ServerCore/Helpers/SubmissionEvaluator.cs

```diff
@@ -0,0 +1,356 @@
+            {
+                return new SubmissionResponse() { ResponseCode = SubmissionResponseCode.TeamLockedOut };
+            }
+
+            duplicateSubmission = (from sub in submissions
+                                   where sub.SubmissionText == Response.FormatSubmission(submissionText)
+                                   select sub).Any();
+
+            if (duplicateSubmission)
+            {
+                return new SubmissionResponse() { ResponseCode = SubmissionResponseCode.DuplicateSubmission };
+            }
+
+            // Create submission and add it to list
+            Submission submission = new Submission
+            {
+                TimeSubmitted = DateTime.UtcNow,
+                Puzzle = puzzleState.Puzzle,
+                Team = puzzleState.Team,
+                Submitter = loggedInUser,
+                AllowFreeformSharing = allowFreeformSharing
+            };
+
+            string submissionTextToCheck = Response.FormatSubmission(submissionText);
+
+            if (puzzle.IsFreeform)
+            {
+                submission.UnformattedSubmissionText = submissionText;
+            }
+            else
+            {
+                submission.SubmissionText = submissionText;
+            }
+
+            submission.Response = await context.Responses.Where(
+                r => r.Puzzle.ID == puzzleId &&
+                     submissionTextToCheck == r.SubmittedText)
+                .FirstOrDefaultAsync();
+
+            submissions.Add(submission);
+
+            // Update puzzle state if submission was correct
+            if (submission.Response != null && submission.Response.IsSolution)
+            {
+                await PuzzleStateHelper.SetSolveStateAsync(context,
+                    thisEvent,
+                    submission.Puzzle,
+                    submission.Team,
+                    submission.TimeSubmitted);
+            }
+            else if (!puzzle.IsFreeform && submission.Response == null && thisEvent.IsAnswerSubmissionActive)
+            {
+                // We also determine if the puzzle should be set to email-only mode.
+                if (IsPuzzleSubmissionLimitReached(
+                        thisEvent,
+                        submissions,
+                        puzzleState))
+                {
+                    await PuzzleStateHelper.SetEmailOnlyModeAsync(context,
+                        thisEvent,
+                        submission.Puzzle,
+                        submission.Team,
+                        true);
+
+                    var authors = await context.PuzzleAuthors.Where((pa) => pa.Puzzle == submission.Puzzle).Select((pa) => pa.Author.Email).ToListAsync();
+
+                    MailHelper.Singleton.SendPlaintextBcc(authors,
+                        $"{thisEvent.Name}: Team {submission.Team.Name} is in email mode for {submission.Puzzle.Name}",
+                        "");

```

Comment:
Should this return a locked result or is this the last submission permitted before the lock?

Reply:
I don't see anything in the logic that blocks submissions after setting this up, so I think it goes through this time.